cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,329 @@
+                  [&] (gdf_index_type const index) { return left_sync_table.get_column(index); });
+    
+    std::vector<gdf_column*> right_key_cols_vect(key_cols.size());
+    std::transform(key_cols.cbegin(), key_cols.cend(), right_key_cols_vect.begin(),
+                  [&] (gdf_index_type const index) { return right_sync_table.get_column(index); });
+
+    auto left_key_table = device_table::create(left_key_cols_vect.size(), left_key_cols_vect.data());
+    auto right_key_table = device_table::create(right_key_cols_vect.size(), right_key_cols_vect.data());
+    rmm::device_vector<int8_t> asc_desc_d(asc_desc);
+
+    rmm::device_vector<gdf_size_type> merged_table_indices;
+    rmm::device_vector<gdf_size_type> merged_row_indices;
+    std::tie(merged_table_indices, merged_row_indices) = generate_merged_indices(*left_key_table, *right_key_table, asc_desc_d, nulls_are_smallest, stream);
+
+    // Allocate output table
+    bool nullable = has_nulls(left_sync_table) || has_nulls(right_sync_table);
+    table destination_table(left_sync_table.num_rows() + right_sync_table.num_rows(), column_dtypes(left_sync_table), nullable, false, stream);
+    for (gdf_size_type i = 0; i < destination_table.num_columns(); i++) {
+        gdf_column const* left_col = left_sync_table.get_column(i);
+        gdf_column * out_col = destination_table.get_column(i);
+        
+        if (left_col->dtype != GDF_STRING_CATEGORY){
+            continue;
+        }
+
+        NVCategory * category = static_cast<NVCategory*>(left_col->dtype_info.category);
+        out_col->dtype_info.category = category->copy();
+    }
+    
+    // Materialize
+    auto left_device_table_ptr = device_table::create(left_sync_table, stream);
+    auto right_device_table_ptr = device_table::create(right_sync_table, stream);
+    auto output_device_table_ptr = device_table::create(destination_table, stream);
+    auto& left_device_table = *left_device_table_ptr;
+    auto& right_device_table = *right_device_table_ptr;
+    auto& output_device_table = *output_device_table_ptr;
+
+    auto index_start_it = thrust::make_zip_iterator(thrust::make_tuple(
+                                                    thrust::make_counting_iterator(static_cast<gdf_size_type>(0)), 
+                                                    merged_table_indices.begin(),
+                                                    merged_row_indices.begin()));
+    auto index_end_it = thrust::make_zip_iterator(thrust::make_tuple(
+                                                thrust::make_counting_iterator(static_cast<gdf_size_type>(merged_table_indices.size())),
+                                                merged_table_indices.end(),
+                                                merged_row_indices.end()));
+
+    thrust::for_each(rmm::exec_policy(stream)->on(stream),
+                    index_start_it,
+                    index_end_it,
+                    [=] __device__ (auto const & idx_tuple){
```

Comment:
Passing tuples as input parameters can be tricky to read and understand. It helps to break up a tuple into named variables. For example:
```
[=] __device__ (auto const & idx_tuple){
   thing0 = thrust::get<0>(idx_tuple);
   thing1 = thrust::get<1>(idx_tuple);
   thing2 = thrust::get<2>(idx_tuple);
```